src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -2314,8 +2364,21 @@ QueryPushdownTaskCreate(Query *originalQuery, int shardIndex,
 			/* use the shard from a specific index */
 			shardInterval = cacheEntry->sortedShardIntervalArray[shardIndex];
 
-			/* use a shard from a distributed table as the anchor shard */
-			anchorShardId = shardInterval->shardId;
+			/*
+			 * We want to use result relation's shard id as anchor shard id
+			 */
+			if (modifyWithSubselect)
+			{
+				if (relationId == resultRelationOid)
+				{
+					anchorShardId = shardInterval->shardId;
+				}
+			}
+			else
+			{
+				/* use a shard from a distributed table as the anchor shard */
+				anchorShardId = shardInterval->shardId;
+			}
```

Comment:
the nesting here is a bit confusing, we could maybe make the code a bit more self-documenting
```c
else if (UpdateOrDeleteQuery(originalQuery))
{
    if (!modifyWithSubquery || relationId == resultRelationOid)
    {
        /* for UPDATE/DELETE the shard in the result relation becomes the anchor shard */
        anchorShardId = shardInterval->shardId;
    }
}
else 
{
    /* for SELECT we pick an arbitrary shard as the anchor shard */
    anchorShardId = shardInterval->shardId;
}
```
could also introduce a boolean variable to avoid calling `UpdateOrDeleteQuery`

Reply:
Could also consider always initialising the `resultRelationOid` for `UpdateOrDeleteQuery` to further improve readability.